File: testers/avro/java/generated-and-checked-in/com/example/events/header/StandardHeaders.java
package com.example.events.header;

import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Optional;
import java.util.UUID;
import org.apache.kafka.common.header.Headers;
import org.apache.kafka.common.header.internals.RecordHeaders;

/** Typed headers for Kafka messages */
public record StandardHeaders(UUID correlationId, Instant timestamp, Optional<String> source) {
  public StandardHeaders withCorrelationId(UUID correlationId) {
    return new StandardHeaders(correlationId, timestamp, source);
  }

  public StandardHeaders withTimestamp(Instant timestamp) {
    return new StandardHeaders(correlationId, timestamp, source);
  }

  public StandardHeaders withSource(Optional<String> source) {
    return new StandardHeaders(correlationId, timestamp, source);
  }

  /** Parse from Kafka Headers */
  public static StandardHeaders fromHeaders(Headers headers) {
    UUID correlationId =
        UUID.fromString(
            new String(headers.lastHeader("correlationId").value(), StandardCharsets.UTF_8));
    Instant timestamp =
        Instant.ofEpochMilli(
            Long.parseLong(
                new String(headers.lastHeader("timestamp").value(), StandardCharsets.UTF_8)));
    Optional<String> source =
        Optional.ofNullable(headers.lastHeader("source"))
            .map(h -> new String(h.value(), StandardCharsets.UTF_8));
    return new StandardHeaders(correlationId, timestamp, source);
  }

  /** Convert to Kafka Headers */
  public Headers toHeaders() {
    Headers headers = new RecordHeaders();
    headers.add("correlationId", correlationId.toString().getBytes(StandardCharsets.UTF_8));
    headers.add(
        "timestamp", Long.toString(timestamp.toEpochMilli()).getBytes(StandardCharsets.UTF_8));
    source.ifPresent(v -> headers.add("source", v.getBytes(StandardCharsets.UTF_8)));
    return headers;
  }
}

File: testers/avro/scala/generated-and-checked-in/com/example/events/header/StandardHeaders.scala
package com.example.events.header

import java.nio.charset.StandardCharsets
import java.time.Instant
import java.util.UUID
import org.apache.kafka.common.header.Headers
import org.apache.kafka.common.header.internals.RecordHeaders

/** Typed headers for Kafka messages */
case class StandardHeaders(
  correlationId: UUID,
  timestamp: Instant,
  source: Option[String]
) {
  /** Convert to Kafka Headers */
  def toHeaders: Headers = {
    val headers: Headers = new RecordHeaders()
    headers.add("correlationId", correlationId.toString().getBytes(StandardCharsets.UTF_8))
    headers.add("timestamp", java.lang.Long.toString(timestamp.toEpochMilli).getBytes(StandardCharsets.UTF_8))
    source.foreach(v => headers.add("source", v.getBytes(StandardCharsets.UTF_8)))
    return headers
  }
}

object StandardHeaders {
  /** Parse from Kafka Headers */
  def fromHeaders(headers: Headers): StandardHeaders = {
    val correlationId: UUID = UUID.fromString(new String(headers.lastHeader("correlationId").value(), StandardCharsets.UTF_8))
    val timestamp: Instant = Instant.ofEpochMilli(java.lang.Long.parseLong(new String(headers.lastHeader("timestamp").value(), StandardCharsets.UTF_8)))
    val source: Option[String] = Option.apply(headers.lastHeader("source")).map(h => new String(h.value(), StandardCharsets.UTF_8))
    return new StandardHeaders(correlationId, timestamp, source)
  }
}